### PR TITLE
[CMAKE] Enable FatLTO as a build option for LLVM

### PR DESCRIPTION
**Pull request by ilovepi**
Since LLVM supports `-ffat-lto-objects` we should enable this as an option in the LLVM build. FatLTO should improve the time it takes to build tests for LTO enabled builds of the compiler by not linking w/ the bitcode portion of the object files, which should speed up build times for LTO builds without disabling optimizations.

### CODE REVIEW

**Comment by ilovepi** [comment]
I'm keeping this as a draft until I confirm the cmake config works as expected, especially in 2 stage builds.

**Comment by ilovepi** [comment]
This speeds up a 2 stage build by about 3 minutes, tests time are still much closer than expected. Locally I saw a bout a 5-8% speedup to build clang unit tests, but maybe the unit tests are too small in the overall test regime to make a difference.

I'll try to collect more performance numbers locally w.r.t. the test times.

File: llvm/cmake/modules/HandleLLVMOptions.cmake [hunk 1]
```diff
@@ -1251,6 +1253,10 @@ elseif(LLVM_ENABLE_LTO)
   endif()
 endif()
 
+if(LLVM_ENABLE_FATLTO AND (FUCHSIA OR UNIX))
+    append("-ffat-lto-objects" CMAKE_EXE_LINKER_FLAGS CMAKE_SHARED_LINKER_FLAGS)
```
[the remark this answers is not in the document]
Comment (ilovepi) [reply llvm/cmake/modules/HandleLLVMOptions.cmake:1257]: Dang, I thought I had done that. That's probably why I'm seeing an odd result. Thank you.

[the remark this answers is not in the document]
Comment (petrhosek) [reply llvm/cmake/modules/HandleLLVMOptions.cmake:1266]: I'd also include the flag in `CMAKE_MODULE_LINKER_FLAGS` for completeness.

File: llvm/cmake/modules/AddLLVM.cmake [hunk 2]
```diff
@@ -1621,8 +1621,15 @@ function(add_unittest test_suite test_name)
   # The runtime benefits of LTO don't outweight the compile time costs for tests.
   if(LLVM_ENABLE_LTO)
     if((UNIX OR MINGW) AND LINKER_IS_LLD)
-      set_property(TARGET ${test_name} APPEND_STRING PROPERTY
-                    LINK_FLAGS " -Wl,--lto-O0")
+      if(LLVM_ENABLE_FATLTO)
+        # When using FatLTO, just use relocatable linking.
+        set_property(TARGET ${test_name} APPEND_STRING PROPERTY
+                      LINK_FLAGS " -Wl,--no-fat-lto-objects")
+        set_property(TARGET ${test_name} APPEND_STRING PROPERTY COMPILE_FLAGS " -fno-lto")
```
Comment (petrhosek) [inline llvm/cmake/modules/AddLLVM.cmake:1628]: What's the motivation for including `-fno-lto` in compile flags here?

Comment (ilovepi) [reply llvm/cmake/modules/AddLLVM.cmake:1628]: This makes sure we use the non-LTO part of the build. It's probably pedantic to add them both, but it makes sure we choose the object file part and not the bitcode section.

File: llvm/cmake/modules/HandleLLVMOptions.cmake (hunk 1 above)
[the remark this answers is not in the document]
Comment (ilovepi) [reply llvm/cmake/modules/HandleLLVMOptions.cmake]: done. Thanks for the suggestion.

File: llvm/cmake/modules/AddLLVM.cmake (hunk 2 above)
Comment (petrhosek) [reply llvm/cmake/modules/AddLLVM.cmake:1628]: I'd omit it since it shouldn't be necessary.

**Review by petrhosek — APPROVED** [review]

**Comment by ilovepi** [comment]
Not sure what I've managed to do here w/ basic rebase ... Will remove  unrelated folks and triage the commits

**Pull request merged**